### PR TITLE
PP-5297: POST /v1/agreements -> POST /v1/directdebit/mandates

### PR DESCRIPTION
**Pull request by oswaldquek**
Also added description to the create mandate request.

## WHAT YOU DID
_A brief description of the pull request:_

## How to test

- How should it be reviewed? 
- What feedback do you need? 
- If manual testing is needed, give suggested testing steps.

### Logging

- [ ] only emit log lines at ERROR level which require immediate attention from a support engineer. These will trigger a zendesk alert.

### Documentation

- [ ] Updated README.md for any of the following ?

* Introduced any new environment variables / removed existing environment variable
* Added new API / updated existing API definition

### CODE REVIEW

File: src/main/java/uk/gov/pay/api/model/PaymentProvider.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.pay.api.model;
+
+public enum PaymentProvider {
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/api/model/PaymentProvider.java]: Do we need this? Won't we just pass on the string returned by dd-connector. Or is this considered better?

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/AgreementStatus.java [hunk 2]
```diff
@@ -1,11 +1,18 @@
 package uk.gov.pay.api.model.directdebit.agreement;
 
+import com.fasterxml.jackson.annotation.JsonFormat;
+
+@JsonFormat(shape = JsonFormat.Shape.OBJECT)
 public enum AgreementStatus {
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/api/model/directdebit/agreement/AgreementStatus.java]: Will this be renamed to MandateStatus?

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/CreateAgreementResponse.java [hunk 3]
```diff
@@ -3,41 +3,48 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
+import uk.gov.pay.api.model.PaymentProvider;
 import uk.gov.pay.api.model.links.directdebit.AgreementLinks;
 
 @ApiModel(value = "CreateAgreementResponse", description = "The Agreement Payload to create a new Agreement")
 public class CreateAgreementResponse {
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/api/model/directdebit/agreement/CreateAgreementResponse.java]: Will this be renamed to CreateMandateResponse?

File: src/main/java/uk/gov/pay/api/resources/AgreementsResource.java [hunk 4]
```diff
@@ -74,7 +74,7 @@ public Response getPayment(@ApiParam(value = "accountId", hidden = true) @Auth A
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/api/resources/AgreementsResource.java]: Will this class be renamed to MandatesResource?

File: src/main/java/uk/gov/pay/api/model/PaymentProvider.java (hunk 1 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/api/model/PaymentProvider.java]: oh at the moment dd-connector doesn't return the payment provider name. i'll need to amend that

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/CreateMandateResponse.java [hunk 5]
```diff
@@ -5,55 +5,64 @@
...
+    private final String paymentProvider;
 
-    private CreateAgreementResponse(String agreementId,
-                                    String providerId,
-                                    String reference,
-                                    String returnUrl,
-                                    String createdDate,
-                                    AgreementStatus state,
-                                    AgreementLinks links) {
+    private CreateMandateResponse(String agreementId,
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/api/model/directdebit/agreement/CreateMandateResponse.java]: mandateId

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java [hunk 6]
```diff
@@ -5,7 +5,7 @@
 import io.swagger.annotations.ApiModelProperty;
 import uk.gov.pay.api.model.links.directdebit.AgreementLinks;
 
-@ApiModel(value = "CreateAgreementResponse", description = "The Agreement Payload to create a new Agreement")
+@ApiModel(value = "CreateMandateResponse", description = "The Agreement Payload to create a new Agreement")
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java]: "The Mandate Payload to create a new Mandate"

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java [hunk 7]
```diff
@@ -64,7 +64,7 @@ public String getReturnUrl() {
 
     @ApiModelProperty(value = "agreement state", required = true, example = "CREATED")
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java]: "mandate state"

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java [hunk 8]
```diff
@@ -5,7 +5,7 @@
 import io.swagger.annotations.ApiModelProperty;
 import uk.gov.pay.api.model.links.directdebit.AgreementLinks;
 
-@ApiModel(value = "CreateAgreementResponse", description = "The Agreement Payload to create a new Agreement")
+@ApiModel(value = "CreateMandateResponse", description = "The Agreement Payload to create a new Agreement")
 public class GetAgreementResponse {
 
     private static final String AGREEMENT_ID_FIELD_NAME = "agreement_id";
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java]: is this still used?

File: src/test/java/uk/gov/pay/api/utils/mocks/ConnectorDDMockClient.java [hunk 9]
```diff
@@ -64,33 +65,27 @@ public void respondWithPaymentCreated(DirectDebitConnectorCreatePaymentResponse
...
-                                serviceReference,
-                                returnUrl,
-                                createdDate,
-                                state,
-                                validGetLink(mandateLocation(gatewayAccountId, mandateId), "self"),
-                                validGetLink(nextUrl(chargeTokenId), "next_url"),
-                                validPostLink(nextUrlPost(), "next_url_post", "application/x-www-form-urlencoded", Map.of("chargeTokenId", chargeTokenId)
+
+    public void respondOk_whenCreateMandateRequest(CreateMandateRequestParams params) {
+        setupCreateAgreement(params.getGatewayAccountId(), aResponse()
```
Comment (stephencdaly) [inline src/test/java/uk/gov/pay/api/utils/mocks/ConnectorDDMockClient.java]: setupCreateMandate

File: src/test/java/uk/gov/pay/api/utils/mocks/ConnectorDDMockClient.java [hunk 10]
```diff
@@ -64,33 +65,27 @@ public void respondWithPaymentCreated(DirectDebitConnectorCreatePaymentResponse
...
+                .withStatus(CREATED_201)
+                .withHeader(CONTENT_TYPE, APPLICATION_JSON)
+                .withHeader(LOCATION, format("/v1/api/accounts/%s/mandates/%s", params.getGatewayAccountId(), params.getMandateId()))
+                .withBody(buildCreateAgreementResponse(
```
Comment (stephencdaly) [inline src/test/java/uk/gov/pay/api/utils/mocks/ConnectorDDMockClient.java]: buildCreateMandateResponse

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java (hunk 6 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java]: i think this should be covered in https://payments-platform.atlassian.net/browse/PP-5299

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java (hunk 7 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java]: i think this should be covered in https://payments-platform.atlassian.net/browse/PP-5299

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java (hunk 8 above)
Comment (oswaldquek) [reply src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java]: i think this should be covered in https://payments-platform.atlassian.net/browse/PP-5299

Comment (stephencdaly) [reply src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java]: Aaah 👍

File: src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java (hunk 6 above)
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/api/model/directdebit/agreement/GetAgreementResponse.java]: 👍

**Pull request merged**